File: DSA-Sheet-Apna-Colllege/Array/6_searchElementInSortedRotatedArray.cpp
// ramnitcode27
// Given a sorted and rotated array A of N distinct elements which is rotated at some point, and given an element key. The task is to find the index of the given element key in the array A.

#include <bits/stdc++.h>
using namespace std;

class Solution{
    public:
    // Time Complexity: O(N)
    int searchN(int a[], int l, int h, int key){
        for(int i=0;i<h-l+1;i++)
            if(a[i]==key)
                return i;
        return -1;
    }
    
    // Time Complexity: O(log N).
    // Auxiliary Space: O(1).
    int search(int arr[], int l, int h, int key){
        if (l > h)
            return -1;
 
        int mid = (l + h) / 2;
        if (arr[mid] == key)
            return mid;
 
        if (arr[l] <= arr[mid]) {
            if (key >= arr[l] && key <= arr[mid])
                return search(arr, l, mid - 1, key);
            return search(arr, mid + 1, h, key);
        }
 
        if (key >= arr[mid] && key <= arr[h])
            return search(arr, mid + 1, h, key);
 
        return search(arr, l, mid - 1, key);
    }
};

int main() {
    int t;
    cin >> t;
    while(t--){
        int n;
        cin >> n;
        int A[n];
        for(int i = 0; i < n; i++)
            cin >> A[i];
        int key;
        cin >> key;
        Solution ob;
        cout << ob.search(A, 0, n - 1, key) << endl;
    }
return 0;
}

File: DSA-Sheet-Apna-Colllege/Array/6_searchElementInSortedRotatedArray.py
# ramnitcode27
# Given a sorted and rotated array A of N distinct elements which is rotated at some point, and given an element key. The task is to find the index of the given element key in the array A.

class Solution:
    # Time Complexity: O(N)
    def searchN(self, A : list, l : int, h : int, key : int):
      for i in range(len(A)):
          if key == A[i]:
              return i
          else:
              continue
      return -1
    
    # Time Complexity: O(log N).
    # Auxiliary Space: O(1).
    def search(self, A : list, l : int, h : int, key : int):
        if l>h:
            return -1
        
        mid=(l+h)//2
        if A[mid]==key:
            return mid
        
        if A[l]<=A[mid]:
            if key >= A[l] and key <= A[mid]:
                return self.search(A, l, mid - 1, key)
            return self.search(A, mid + 1, h, key)
            
        if key >= A[mid] and key <= A[h]:
            return self.search(A, mid + 1, h, key)

        return self.search(A, l, mid - 1, key)

if __name__ == '__main__':
    t = int(input())
    
    for _ in range(t):
        n = int(input())
        A = [int(x) for x in input().split()]
        k = int(input())
        ob=Solution()
        print(ob.search(A, 0, n - 1, k))